File: matlab/tests/test_post_hoc_guards.m
%% test_post_hoc_guards.m — the postHocGuards default and its wiring
%
%  Two checks in the toolbox inspect a route's output after computing it
%  and may then recompute by another route: the nested accuracy guard in
%  nestedContract/combinePair and the corruption check in the flat cosine
%  path. With either active the measured cost of the Mobius route is not
%  the cost of choosing it, because a diverting check pays for both
%  routes. The default switches them off so the routes can be timed as
%  the alternatives they are.
%
%  Python twin: tests/test_post_hoc_guards.py. The Python file also
%  exercises combinePair directly; that function is local to
%  nestedContract.m and unreachable from here, so this file covers the
%  default's registration, validation, and the flat cosine wiring.
%
%  Standalone-runnable; appends to `results` when called from test_mpt.m.

if ~exist('results', 'var')
    results = {};
    standalone = true;
    % Defaults isolation when run standalone (when invoked from
    % test_mpt.m the outer wrapper has already isolated defaults).
    addpath(fileparts(mfilename('fullpath')));
    clear cleanupDefaults
    cleanupDefaults = mptTestIsolateDefaults(); %#ok<NASGU>
else
    standalone = false;
end

% --- 1. Default is on ---
phg_prev = mptDefaults('postHocGuards');
results{end+1, 1} = 'postHocGuards: factory default is true'; %#ok<*AGROW>
results{end, 2}   = islogical(phg_prev) && isscalar(phg_prev) && phg_prev;

% --- 2. Round-trips ---
mptDefaults('postHocGuards', false);
phg_ok = ~mptDefaults('postHocGuards');
mptDefaults('postHocGuards', true);
phg_ok = phg_ok && mptDefaults('postHocGuards');
results{end+1, 1} = 'postHocGuards: set/get round-trips';
results{end, 2}   = phg_ok;

% --- 3. Rejects a non-logical value ---
phg_ok = false;
try
    mptDefaults('postHocGuards', 'yes');
catch phg_err
    phg_ok = strcmp(phg_err.identifier, 'mptDefaults:badValue');
end
results{end+1, 1} = 'postHocGuards: rejects a non-logical value';
results{end, 2}   = phg_ok;

% --- 4. Survives a reset ---
mptDefaults('postHocGuards', false);
mptDefaults('reset');
results{end+1, 1} = 'postHocGuards: reset restores true';
results{end, 2}   = logical(mptDefaults('postHocGuards'));

% --- 5. Flat cosine agrees with the guard off ---
%  The corruption check fires only on a broken value (non-finite,
%  negative auto-IP, or |cosine| > 1), so on a healthy input the two
%  settings must give the same answer bit for bit. Fixed pitches rather
%  than rand(), so this file leaves the global RNG stream untouched for
%  the files that run after it.
phg_P  = 1200;
phg_p1 = [0; 137; 291; 452; 613; 761; 908; 1063];
phg_p2 = [24; 160; 318; 470; 640; 788; 931; 1085];
phg_w  = ones(8, 1);

mptDefaults('postHocGuards', true);
phg_sOn  = cosSimExpTens(phg_p1, phg_w, phg_p2, phg_w, 30, 3, ...
                         false, true, phg_P, 'verbose', false);
mptDefaults('postHocGuards', false);
phg_sOff = cosSimExpTens(phg_p1, phg_w, phg_p2, phg_w, 30, 3, ...
                         false, true, phg_P, 'verbose', false);
mptDefaults('postHocGuards', phg_prev);

results{end+1, 1} = 'postHocGuards: flat cosine unchanged on a healthy input';
results{end, 2}   = isequal(phg_sOn, phg_sOff);

if standalone
    nPass = sum(cellfun(@(x) isequal(x, true), results(:,2)));
    nFail = numel(results(:,1)) - nPass;
    fprintf('\n=== test_post_hoc_guards: %d passed, %d failed (of %d) ===\n', ...
        nPass, nFail, numel(results(:,1)));
    % Restore caller's pre-test defaults eagerly
    % (fires the helper's onCleanup destructor before
    % the conditional error below halts execution).
    clear cleanupDefaults
    if nFail > 0
        for ii = 1:size(results, 1)
            if ~isequal(results{ii, 2}, true)
                fprintf('  FAIL  %s\n', results{ii, 1});
            end
        end
    end
end

File: python/tests/test_post_hoc_guards.py
"""The ``post_hoc_guards`` default and what switching it off does.

Two checks in the toolbox inspect a route's output after computing it
and may then recompute by another route: the nested accuracy guard in
``_combine_pair`` and the corruption check in the flat cosine path. With
either active the measured cost of the Möbius route is not the cost of
choosing it, because a diverting check pays for both routes. The default
switches them off so the routes can be timed as the alternatives they
are.

Mirror of MATLAB ``tests/test_post_hoc_guards.m``.
"""
import numpy as np
import pytest

import mpt
from mpt._defaults import truncation_floor
from mpt._tensor._nested_contraction import (
    _combine_pair, _tuple_sides, _combine_chunked, _combine_orbit,
    _ORBIT_ENUM_MAX_ELEMS, orbit_guard_scope,
)


def test_default_is_on_and_round_trips():
    assert mpt.get_defaults()["post_hoc_guards"] is True
    try:
        mpt.set_default(post_hoc_guards=False)
        assert mpt.get_defaults()["post_hoc_guards"] is False
    finally:
        mpt.reset_defaults()
    assert mpt.get_defaults()["post_hoc_guards"] is True


def test_rejects_non_boolean():
    with pytest.raises(ValueError, match="post_hoc_guards"):
        mpt.set_default(post_hoc_guards="yes")
    mpt.reset_defaults()


def _block(r, K, seed=0):
    return np.random.default_rng(seed).uniform(0.0, 1.0, (8, K, K))


def test_guard_on_diverts_to_enumeration_at_r6():
    """At r = 6 the bound exceeds the floor, so the guard should divert:
    the returned value is enumeration's, not the Möbius route's."""
    r, K = 6, 6
    M = _block(r, K)
    xt, yt = _tuple_sides(K, r, True)
    enum = np.asarray(_combine_chunked(M, xt, yt, _ORBIT_ENUM_MAX_ELEMS))
    orbit, bound = _combine_orbit(M, r, return_bound=True)
    assert float(np.max(bound)) > truncation_floor(6), (
        "premise: the guard must fire at this shape"
    )
    assert not np.array_equal(np.asarray(orbit), enum), (
        "premise: the two routes must differ for this test to bite"
    )
    try:
        mpt.set_default(post_hoc_guards=True)
        with orbit_guard_scope(6.0):
            got = np.asarray(_combine_pair(M, r, True, True))
    finally:
        mpt.reset_defaults()
    np.testing.assert_array_equal(got, enum)


def test_guard_off_keeps_the_mobius_result_at_r6():
    """With the guard off the same block returns the Möbius value."""
    r, K = 6, 6
    M = _block(r, K)
    orbit = np.asarray(_combine_orbit(M, r, return_bound=False))
    try:
        mpt.set_default(post_hoc_guards=False)
        with orbit_guard_scope(6.0):
            got = np.asarray(_combine_pair(M, r, True, True))
    finally:
        mpt.reset_defaults()
    np.testing.assert_array_equal(got, orbit)


def test_guard_off_still_within_the_stated_accuracy_here():
    """Switching the guard off is not a licence to exceed the accuracy
    truncationSigmas states: at this shape the bound is loose and the
    Möbius route's actual error sits well inside the floor."""
    r, K = 6, 6
    M = _block(r, K)
    xt, yt = _tuple_sides(K, r, True)
    enum = np.asarray(_combine_chunked(M, xt, yt, _ORBIT_ENUM_MAX_ELEMS))
    orbit = np.asarray(_combine_orbit(M, r, return_bound=False))
    assert np.max(np.abs(orbit - enum)) < truncation_floor(6)
